src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/CSharpTagHelperCodeRenderer.cs

```diff
@@ -340,18 +332,72 @@ private void RenderUnboundHTMLAttributes(IEnumerable<KeyValuePair<string, Chunk>
             }
         }
 
-        private void RenderTagHelperBody(IList<Chunk> children, bool bufferBody)
+        private void RenderTagHelperContent()
         {
-            // If we want to buffer the body we need to create a writing scope to capture the body content.
-            if (bufferBody)
+            // Rendering output is a runtime feature.
+            if (_designTimeMode)
             {
-                // Render all of the tag helper children in a buffered writing scope.
-                BuildBufferedWritingScope(children);
+                return;
             }
-            else
+
+            _writer.Write("if (")
+                   .Write(ExecutionContextVariableName)
+                   .Write(".")
+                   .Write(_tagHelperContext.ExecutionContextOutputPropertyName)
+                   .Write(".")
+                   .Write(_tagHelperContext.OutputContentSetPropertyName)
+                   .WriteLine(")");
+
+            // Render the body of the if statement, at this point in the codegen the TagHelperOutput's Content was set
+            // so we should just render its Content, no need to forcefully execute the child content.
+            using (_writer.BuildScope())
             {
-                // Render all of the tag helper children.
-                _bodyVisitor.Accept(children);
+                RenderTagOutput(_tagHelperContext.OutputGenerateContentMethodName);
+            }
+
+            _writer.Write("else if (")
+                   .Write(ExecutionContextVariableName)
+                   .Write(".")
+                   .Write(_tagHelperContext.ExecutionContextChildContentRetrievedPropertyName)
+                   .WriteLine(")");
+
+            // Render the body of the else if statement, at this point in the codegen the GetChildContentAsync method 
+            // was invoked but the TagHelperOutput's Content was not set. Call into GetChildContentAsync to retrieve
+            // the cached value of the content so we don't execute the child content twice.
```

Comment:
also, if the content wasn't set but `GetChildContentAsync` was called, doesn't this block override the intention of the tag helper(s) author(s) involved?  someone decided to make the call and _not_ to use the returned information.